src/omd.mli

```diff
@@ -36,29 +30,23 @@ type list_spacing =
   | Loose
   | Tight
 
-type def_elt =
+type 'a def_elt =
   {
-    term: inline;
-    defs: inline list;
+    term: 'a;
+    defs: 'a list;
   }
 
-and block =
-  {
-    bl_desc: block_desc;
-    bl_attributes: attributes;
-  }
-
-and block_desc =
-  | Paragraph of inline
-  | List of list_type * list_spacing * block list list
-  | Blockquote of block list
-  | Thematic_break
-  | Heading of int * inline
-  | Code_block of string * string
-  | Html_block of string
-  | Definition_list of def_elt list
-
-type doc = block list
+type ('attr, 'a) block =
+  | Paragraph of 'attr * 'a
+  | List of 'attr * list_type * list_spacing * ('attr, 'a) block list list
+  | Blockquote of 'attr * ('attr, 'a) block list
+  | Thematic_break of 'attr
+  | Heading of 'attr * int * 'a
+  | Code_block of 'attr * string * string
+  | Html_block of 'attr * string
+  | Definition_list of 'attr * 'a def_elt list
+
+type doc = (attributes, attributes inline) block list
```

Comment:
How about exposing the `Ast.t` in this interface and then using it here?

```suggestion
type doc = attributes Ast.t list
```

Reply:
This is the thing that is making me consider moving back to using a functor as we were using before in the Ast module. There's no way to expose block without a parameter here. It feels weird to have `Omd.block`, ``Omd.t` and `Omd.doc`.

I'd like to either

1. Pick a different name for `block` and use the name `block` for `t` and update the `Ast` module to match it
2. Go back to having the functor in the `Ast` module as it was before so that `block` doesn't have a type parameter and we don't need a `t`.

Let me know what you think and I'll try to wrap up this up by tomorrow so that we can unblock the other things.

Reply:
I'm afraid I don't quite follow. I think it makes sense to have a these two different type constructors exposed:

- `('attr, 'a) block`: types for block-level elements of any attributes and any content  
- `'attr Ast.t`: types for the standard markdown AST, over any attributes 

And to have this additional type:

- `doc = attributes t list`: the type for the standard representation of a markdown document, with the standard attributes. 

Tho I could also see an argument `Ast.t` being defines as a list of blocks.
(Well, if I were designing from scratch, I'd probably call `doc` `Omd.t`, but that's inessential.)

That said, if you think some use of simple functors will improve the interface and make a nicer presentation of types, I wouldn't be opposed. Do you want to sketch out what you have in mind?

Reply:
The `Ast` module used a [functor](https://github.com/ocaml/omd/blob/master/src/ast.ml#L30). I'm thinking of reintroducing that so that `Omd` can have the following signature:

```
(* ...  *)

type 'attr inline =
  | Concat of 'attr * 'attr inline list
  | Text of 'attr * string
  | Emph of 'attr * 'attr inline
  | Strong of 'attr * 'attr inline
  | Code of 'attr * string
  | Hard_break of 'attr
  | Soft_break of 'attr
  | Link of 'attr * 'attr inline link
  | Image of 'attr * 'attr inline link
  | Html of 'attr * string

(* ... *)

type 'attr block =
  | Paragraph of 'attr * 'attr inline
  | List of 'attr * list_type * list_spacing * 'attr block list list
  | Blockquote of 'attr * 'attr block list
  | Thematic_break of 'attr
  | Heading of 'attr * int * 'attr inline
  | Code_block of 'attr * string * string
  | Html_block of 'attr * string
  | Definition_list of 'attr * 'attr def_elt list

type doc = attributes block list
```

With the current setup in my branch, we can't really expose Ast.t in Omd without exposing the Ast module as well (maybe with a restricted signature).  I find that this Omd signature, the solution currently in my branch, is awkward:

```
(* ... *)
type ('attr, 'a) block =
  | Paragraph of 'attr * 'a
  | List of 'attr * list_type * list_spacing * ('attr, 'a) block list list
  | Blockquote of 'attr * ('attr, 'a) block list
  | Thematic_break of 'attr
  | Heading of 'attr * int * 'a
  | Code_block of 'attr * string * string
  | Html_block of 'attr * string
  | Definition_list of 'attr * 'a def_elt list

type 'attr t = ('attr, 'attr inline) block (* equivalent to Ast.t *)

type doc = attributes t list
(** A markdown document *)
```

I would either change the name of `block` to something else and change `t` to `block` or change `t` to something else.

Reply:
Added back the functor to the `Ast` module. There's no real cons internally besides a heavier weight syntax. I prefer the simpler external interface with a single parameter for the attributes on the `block` and `inline` types in `Omd`.

If you give me your approval, I'll squash the commits and merge this.